# SkillFactory
## Введение в ML, введение в sklearn

В этом задании мы с вами рассмотрим данные с конкурса [Задача предсказания отклика клиентов ОТП Банка](http://www.machinelearning.ru/wiki/index.php?title=%D0%97%D0%B0%D0%B4%D0%B0%D1%87%D0%B0_%D0%BF%D1%80%D0%B5%D0%B4%D1%81%D0%BA%D0%B0%D0%B7%D0%B0%D0%BD%D0%B8%D1%8F_%D0%BE%D1%82%D0%BA%D0%BB%D0%B8%D0%BA%D0%B0_%D0%BA%D0%BB%D0%B8%D0%B5%D0%BD%D1%82%D0%BE%D0%B2_%D0%9E%D0%A2%D0%9F_%D0%91%D0%B0%D0%BD%D0%BA%D0%B0_%28%D0%BA%D0%BE%D0%BD%D0%BA%D1%83%D1%80%D1%81%29)

In [9]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

%matplotlib inline

plt.style.use('ggplot')
plt.rcParams['figure.figsize'] = (12,5)

### Грузим данные

Считаем описание данных

In [10]:

df_descr = pd.read_csv('data/otp_description.csv', sep='\t', encoding='utf8')

In [11]:
for i in range(df_descr.shape[0]):
    print(df_descr.iloc[i,0], "-", df_descr.iloc[i,1])

AGREEMENT_RK - уникальный идентификатор объекта в выборке
AGE - возраст клиента
SOCSTATUS_WORK_FL - социальный статус клиента относительно работы (1 - работает, 0 - не работает)
SOCSTATUS_PENS_FL - социальный статус клиента относительно пенсии (1 - пенсионер, 0 - не пенсионер)
GENDER - пол клиента (1- женщины, 0-мужчины)
CHILD_TOTAL - количество детей клиента
DEPENDANTS - количество иждивенцев клиента
EDUCATION - образование
MARITAL_STATUS - семейное положение
GEN_INDUSTRY - отрасль работы клиента
GEN_TITLE - должность
ORG_TP_STATE - форма собственности компании
ORG_TP_FCAPITAL - отношение к иностранному капиталу
JOB_DIR - направление деятельности в нутри компании
FAMILY_INCOME - семейный доход (несколько категорий)
PERSONAL_INCOME - личный доход клиента (в рублях)
REG_ADDRESS_PROVINCE - область регистрации клиента
FACT_ADDRESS_PROVINCE - область фактического пребывания клиента
POSTAL_ADDRESS_PROVINCE - почтовый адрес область
TP_PROVINCE - область торговой точки, где клиент брал послед

Считаем обучающую выборки и тестовую (которую мы как бы не видим)

In [12]:
df_train = pd.read_csv('data/otp_train.csv', sep='\t', encoding='utf8')

In [13]:
df_train.shape

(15223, 52)

In [14]:
df_test = pd.read_csv('data/otp_test.csv', sep='\t', encoding='utf8')

In [15]:
df_test.shape

(14910, 52)

In [16]:
df_train.head()

AGREEMENT_RK  TARGET  AGE  SOCSTATUS_WORK_FL  SOCSTATUS_PENS_FL  GENDER  \
0      59910150       0   49                  1                  0       1   
1      59910230       0   32                  1                  0       1   
2      59910525       0   52                  1                  0       1   
3      59910803       0   39                  1                  0       1   
4      59911781       0   30                  1                  0       0   

   CHILD_TOTAL  DEPENDANTS            EDUCATION  MARITAL_STATUS  \
0            2           1  Среднее специальное  Состою в браке   
1            3           3              Среднее  Состою в браке   
2            4           0     Неполное среднее  Состою в браке   
3            1           1               Высшее  Состою в браке   
4            0           0              Среднее  Состою в браке   

             ...             REG_PHONE_FL GEN_PHONE_FL LOAN_NUM_TOTAL  \
0            ...                        0            1              1   
1            ...                        0            1              1   
2            ...                        0            1              2   
3            ...                        1            1              1   
4            ...                        0            1              2   

  LOAN_NUM_CLOSED LOAN_NUM_PAYM LOAN_DLQ_NUM LOAN_MAX_DLQ  LOAN_AVG_DLQ_AMT  \
0               1             6            2            1              1580   
1               1             6            1            1              4020   
2               1            11            0            0                 0   
3               1             6            3            1  1589,92333333333   
4               1            16            2            1           1152,15   

  LOAN_MAX_DLQ_AMT PREVIOUS_CARD_NUM_UTILIZED  
0             1580                        NaN  
1             4020                        NaN  
2                0                        NaN  
3             1590                        NaN  
4             2230                        NaN  

[5 rows x 52 columns]

## Объединим две выборки

Так как пока мы пока не умеем работать sklearn  Pipeline, то для того, чтобы после предобработки столбцы в двух выборках находились на своих местах.

Для того, чтобы в дальнейшем отделить их введем новый столбец "sample"

In [17]:
df_train.loc[:, 'sample'] = 'train'
df_test.loc[:, 'sample'] = 'test'

In [18]:
df = df_test.append(df_train).reset_index(drop=True)

In [19]:
df.shape

(30133, 53)

### Чуть-чуть посмотрим на данные

Посмотрим типы данных и их заполняемость

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30133 entries, 0 to 30132
Data columns (total 53 columns):
AGREEMENT_RK                  30133 non-null int64
TARGET                        30133 non-null int64
AGE                           30133 non-null int64
SOCSTATUS_WORK_FL             30133 non-null int64
SOCSTATUS_PENS_FL             30133 non-null int64
GENDER                        30133 non-null int64
CHILD_TOTAL                   30133 non-null int64
DEPENDANTS                    30133 non-null int64
EDUCATION                     30133 non-null object
MARITAL_STATUS                30133 non-null object
GEN_INDUSTRY                  27420 non-null object
GEN_TITLE                     27420 non-null object
ORG_TP_STATE                  27420 non-null object
ORG_TP_FCAPITAL               27425 non-null object
JOB_DIR                       27420 non-null object
FAMILY_INCOME                 30133 non-null object
PERSONAL_INCOME               30133 non-null object
REG_ADDRESS_PRO

Видим, что часть данных - object, скорее всего стоки.


Давайте выведем эти значения для каждого столбца

In [21]:
for i in df_train.columns: # перебираем все столбцы
    if str(df_train[i].dtype) == 'object': # если тип столбца - object
        print('='*10)
        print(i) # выводим название столбца
        print(set(df_train[i])) # выводим все его значения (но делаем set - чтоб значения не повторялись)
        print('\n') # выводим пустую строку

EDUCATION
{'Два и более высших образования', 'Неоконченное высшее', 'Высшее', 'Ученая степень', 'Неполное среднее', 'Среднее специальное', 'Среднее'}


MARITAL_STATUS
{'Гражданский брак', 'Разведен(а)', 'Вдовец/Вдова', 'Не состоял в браке', 'Состою в браке'}


GEN_INDUSTRY
{nan, 'Химия/Парфюмерия/Фармацевтика', 'Маркетинг', 'Управляющая компания', 'Информационные услуги', 'СМИ/Реклама/PR-агенства', 'Недвижимость', 'Строительство', 'Ресторанный бизнес/Общественное питание', 'Сельское хозяйство', 'Салоны красоты и здоровья', 'Транспорт', 'Торговля', 'Страхование', 'Нефтегазовая промышленность', 'ЧОП/Детективная д-ть', 'Наука', 'Металлургия/Промышленность/Машиностроение', 'Подбор персонала', 'Коммунальное хоз-во/Дорожные службы', 'Энергетика', 'Государственная служба', 'Информационные технологии', 'Туризм', 'Другие сферы', 'Логистика', 'Юридические услуги/нотариальные услуги', 'Здравоохранение', 'Образование', 'Развлечения/Искусство', 'Банк/Финансы', 'Сборочные производства'}


GEN_TITLE


Mожно заметить что некоторые переменные, которые обозначены как строки (например PERSONAL_INCOME) на самом деле числа, но по какой-то причине были распознаны как строки

Причина же что использовалась запятая для разделения не целой части числа..

Перекодировать их можно например так:

In [22]:
df['PERSONAL_INCOME'].map(lambda x: x.replace(',', '.')).astype('float')

0         7000.0
1         4100.0
2        40000.0
3        20000.0
4        17000.0
5         7000.0
6        10000.0
7         7500.0
8         9000.0
9         4000.0
10       15000.0
11       12000.0
12        8000.0
13       20000.0
14       18500.0
15       14000.0
16        6000.0
17        9000.0
18        5000.0
19       15000.0
20        6100.0
21       22180.0
22        9000.0
23        8700.0
24        7000.0
25        6000.0
26       15000.0
27       12000.0
28       11000.0
29        8000.0
          ...   
30103    12000.0
30104    10000.0
30105    10000.0
30106     8000.0
30107    11000.0
30108     7000.0
30109    10000.0
30110     8000.0
30111    25000.0
30112    10000.0
30113    12000.0
30114     7000.0
30115    12000.0
30116    14500.0
30117    12000.0
30118     8000.0
30119     7000.0
30120    10000.0
30121    10000.0
30122    10000.0
30123    15000.0
30124     8100.0
30125     9500.0
30126    13000.0
30127    10000.0
30128    12000.0
30129    10000.0
30130     6000

Такой эффект наблюдается в столбцах `PERSONAL_INCOME`, `CREDIT`, `FST_PAYMENT`, `LOAN_AVG_DLQ_AMT`, `LOAN_MAX_DLQ_AMT`

### Теперь ваше небольшое исследование

#### Задание 1. Есть ли пропуски в данных? Что с ними сделать?

(единственного верного ответа нет - аргументируйте)

In [23]:
#посчитаем пропуски в данных:
with_null = df.isnull().sum(axis = 0)
with_null[with_null != 0]

GEN_INDUSTRY                   2713
GEN_TITLE                      2713
ORG_TP_STATE                   2713
ORG_TP_FCAPITAL                2708
JOB_DIR                        2713
TP_PROVINCE                     590
REGION_NM                         2
WORK_TIME                      2717
PREVIOUS_CARD_NUM_UTILIZED    29533
dtype: int64

_решение:_ пропуски в данных есть, но способы их заполнения могут зависеть от причин.
для этого рассмотрим пропуски по столбцам отдельно
<br><br>
На сайте сказано, что данные пустые, так как клиенты не заполнили эти поля, но в любом случае нужно проверить на зависимость от других параметром

#### a) Факторы с одинаковым количеством пропусков

У факторов 'GEN_INDUSTRY', 'GEN_TITLE', 'ORG_TP_STATE', 'JOB_DIR' одинаковое количество пропусков. Проверим, находятся ли пропуски в одних и тех же строках. Например, есть ли в этих факторах не пустые значения, если значение 'GEN_INDUSTRY' пустое

In [24]:
df[df['GEN_INDUSTRY'].isnull() == True][['GEN_INDUSTRY',
                                         'GEN_TITLE',
                                         'ORG_TP_STATE',
                                         'JOB_DIR']].isnull().sum(axis = 0)

GEN_INDUSTRY    2713
GEN_TITLE       2713
ORG_TP_STATE    2713
JOB_DIR         2713
dtype: int64

In [25]:
#создадим отдельный df и посмотрим его подробнее
simult_na = df[df['GEN_INDUSTRY'].isnull() == True]
simult_na.head(n = 15).T

1                       5    \
AGREEMENT_RK                              59910420                59911769   
TARGET                                           0                       0   
AGE                                             48                      44   
SOCSTATUS_WORK_FL                                0                       0   
SOCSTATUS_PENS_FL                                1                       1   
GENDER                                           1                       0   
CHILD_TOTAL                                      0                       2   
DEPENDANTS                                       0                       2   
EDUCATION                                  Среднее     Среднее специальное   
MARITAL_STATUS                      Состою в браке          Состою в браке   
GEN_INDUSTRY                                   NaN                     NaN   
GEN_TITLE                                      NaN                     NaN   
ORG_TP_STATE                                   NaN                     NaN   
ORG_TP_FCAPITAL                                NaN                     NaN   
JOB_DIR                                        NaN                     NaN   
FAMILY_INCOME               от 10000 до 20000 руб.  от 10000 до 20000 руб.   
PERSONAL_INCOME                               4100                    7000   
REG_ADDRESS_PROVINCE             Орловская область       Кировская область   
FACT_ADDRESS_PROVINCE            Орловская область       Кировская область   
POSTAL_ADDRESS_PROVINCE          Орловская область       Кировская область   
TP_PROVINCE                      Орловская область       Кировская область   
REGION_NM                            ЦЕНТРАЛЬНЫЙ 1             ПРИВОЛЖСКИЙ   
REG_FACT_FL                                      1                       1   
FACT_POST_FL                                     1                       1   
REG_POST_FL                                      1                       1   
REG_FACT_POST_FL                                 1                       1   
REG_FACT_POST_TP_FL                              1                       1   
FL_PRESENCE_FL                                   0                       0   
OWN_AUTO                                         0                       0   
AUTO_RUS_FL                                      0                       0   
HS_PRESENCE_FL                                   1                       0   
COT_PRESENCE_FL                                  0                       0   
GAR_PRESENCE_FL                                  1                       0   
LAND_PRESENCE_FL                                 0                       0   
CREDIT                                        6000                 13824,2   
TERM                                             6                       5   
FST_PAYMENT                                   2000                   10239   
DL_DOCUMENT_FL                                   0                       0   
GPF_DOCUMENT_FL                                  1                       1   
FACT_LIVING_TERM                               299                      84   
WORK_TIME                                      NaN                     NaN   
FACT_PHONE_FL                                    0                       0   
REG_PHONE_FL                                     0                       0   
GEN_PHONE_FL                                     0                       0   
LOAN_NUM_TOTAL                                   2                       1   
LOAN_NUM_CLOSED                                  1                       1   
LOAN_NUM_PAYM                                   15                       5   
LOAN_DLQ_NUM                                     0                       0   
LOAN_MAX_DLQ                                     0                       0   
LOAN_AVG_DLQ_AMT                                 0                       0   
LOAN_MAX_DLQ_AMT                                 0                   

In [26]:
#скорее всего, данные не заполнены для всех неработающих и/или пенсионеров.
#посчитаем к-во записей по столбцам SOCSTATUS_WORK_FL и SOCSTATUS_PENS_FL в
#этом же df

p = simult_na.pivot_table(
    'TARGET', 'SOCSTATUS_WORK_FL', 'SOCSTATUS_PENS_FL', 'count')
p
#получается, что проблема была в отстутствии данных о работе у неработающих
#пенсионеров. При этом безработные в выборке отсутствуют
#также данных по перечисленным факторам нет для одного 24-летнего.
#Предположим, что это ошибка в данных, тогда пропуск можно заполнить
#самыми частыми значениями для его возрастной группы

SOCSTATUS_PENS_FL    0       1
SOCSTATUS_WORK_FL             
0                  NaN  2712.0
1                  1.0     NaN

#### b) Фактор worktime

Похоже, пропуски в worktime тоже связаны с неработающими пенсионерами из предыдущего пункта. Однако пропусков там чуть больше, найдем в чем разница

In [27]:
no_worktime = df[df['WORK_TIME'].isnull() == True]
o_worktime = pd.merge(no_worktime, simult_na, how='outer')
i_worktime = pd.merge(no_worktime, simult_na)
set_diff_worktime = pd.concat([o_worktime, i_worktime]).drop_duplicates(keep=False)
#set_diff_worktime.T #в df 4 записи
#есть 4 человека со статусом безработный пенсионер, у которых есть данные о работодателе
#при этом не заполнены данные по времени работы в текущем месте. в остальном
#этот фактор совпадает с 4 предыдущими

#### c) Фактор ORG_TP_FCAPITAL

In [28]:
#скорее всего, пропуски в этом факторе тоже связаны с отсутствием данных по безработным пенсионерам
#поэтому найдем разницу между двумя df с пропусками в этих факторах

no_capital = df[df['ORG_TP_FCAPITAL'].isnull() == True]
o_capital = pd.merge(no_capital, simult_na, how='outer')
i_capital = pd.merge(no_capital, simult_na)
set_diff_capital = pd.concat([o_capital, i_capital]).drop_duplicates(keep=False)
#set_diff_capital.T #7 записей в df

Два множества отличаются на 7 записей. Будем считать, что причина одинаковая

#### d) Фактор TP_PROVINCE

TP_PROVINCE = область торговой точки, где клиент брал последний кредит
из описания данных на сайте не совсем понятно, какие клиенты "лежат" в датасете. Значит ли это, что клиент не брал кредит или что данных нет по какой-то другой причине (например, кредит был получен не в магазине)?

In [29]:
#проверим есть ли нули в поле CREDIT - это может означать, что в выборке есть клиенты без кредитов,
#тогда для них не будет и торговой точки выдачи кредита.
df[df['CREDIT'] == 0]

#в поле с суммами выданных кредитов нет пропусков и нулей

Empty DataFrame
Columns: [AGREEMENT_RK, TARGET, AGE, SOCSTATUS_WORK_FL, SOCSTATUS_PENS_FL, GENDER, CHILD_TOTAL, DEPENDANTS, EDUCATION, MARITAL_STATUS, GEN_INDUSTRY, GEN_TITLE, ORG_TP_STATE, ORG_TP_FCAPITAL, JOB_DIR, FAMILY_INCOME, PERSONAL_INCOME, REG_ADDRESS_PROVINCE, FACT_ADDRESS_PROVINCE, POSTAL_ADDRESS_PROVINCE, TP_PROVINCE, REGION_NM, REG_FACT_FL, FACT_POST_FL, REG_POST_FL, REG_FACT_POST_FL, REG_FACT_POST_TP_FL, FL_PRESENCE_FL, OWN_AUTO, AUTO_RUS_FL, HS_PRESENCE_FL, COT_PRESENCE_FL, GAR_PRESENCE_FL, LAND_PRESENCE_FL, CREDIT, TERM, FST_PAYMENT, DL_DOCUMENT_FL, GPF_DOCUMENT_FL, FACT_LIVING_TERM, WORK_TIME, FACT_PHONE_FL, REG_PHONE_FL, GEN_PHONE_FL, LOAN_NUM_TOTAL, LOAN_NUM_CLOSED, LOAN_NUM_PAYM, LOAN_DLQ_NUM, LOAN_MAX_DLQ, LOAN_AVG_DLQ_AMT, LOAN_MAX_DLQ_AMT, PREVIOUS_CARD_NUM_UTILIZED, sample]
Index: []

[0 rows x 53 columns]

In [30]:
#в датафрейме кроме TP_PROVINCE есть еще 4 фактора, относящихся к географии,
#и 4 фактора, указывающие на совпадение этих измерений (среди них есть и
#совпадение по отсутствующему полю)
#посмотрим этот набор более подробно:

geo_col_list = ['REG_ADDRESS_PROVINCE', 'FACT_ADDRESS_PROVINCE', 'POSTAL_ADDRESS_PROVINCE',
               'TP_PROVINCE', 'REGION_NM', 'REG_FACT_FL', 'REG_POST_FL', 'REG_FACT_POST_FL',
               'REG_FACT_POST_TP_FL']

geo_df = df[geo_col_list]
geo_df.head()

REG_ADDRESS_PROVINCE FACT_ADDRESS_PROVINCE POSTAL_ADDRESS_PROVINCE  \
0    Красноярский край     Красноярский край       Красноярский край   
1    Орловская область     Орловская область       Орловская область   
2                 Коми                  Коми                    Коми   
3    Читинская область     Читинская область       Читинская область   
4     Хабаровский край      Хабаровский край        Хабаровский край   

         TP_PROVINCE           REGION_NM  REG_FACT_FL  REG_POST_FL  \
0  Красноярский край  ВОСТОЧНО-СИБИРСКИЙ            0            0   
1  Орловская область       ЦЕНТРАЛЬНЫЙ 1            1            1   
2               Коми       ЦЕНТРАЛЬНЫЙ 2            1            1   
3  Читинская область  ВОСТОЧНО-СИБИРСКИЙ            1            1   
4   Хабаровский край     ДАЛЬНЕВОСТОЧНЫЙ            1            1   

   REG_FACT_POST_FL  REG_FACT_POST_TP_FL  
0                 0                    1  
1                 1                    1  
2                 1                    1  
3                 1                    1  
4                 1                    1

In [31]:
#отдельно посмотрим географические факторы с пропусками в TP_PROVINCE
geo_df = geo_df[geo_df['TP_PROVINCE'].isnull() == True]
geo_df

REG_ADDRESS_PROVINCE FACT_ADDRESS_PROVINCE POSTAL_ADDRESS_PROVINCE  \
40        Московская область    Московская область      Московская область   
67        Московская область    Московская область      Московская область   
120       Московская область    Московская область      Московская область   
282       Московская область    Московская область      Московская область   
315       Московская область    Московская область      Московская область   
336       Московская область    Московская область      Московская область   
534       Московская область    Московская область      Московская область   
537                   Москва                Москва                  Москва   
609                   Москва                Москва                  Москва   
664       Московская область    Московская область      Московская область   
731     Оренбургская область    Московская область      Московская область   
961       Московская область    Московская область      Московская область   
963       Московская область    Московская область      Московская область   
977       Московская область    Московская область      Московская область   
983       Московская область    Московская область      Московская область   
1067      Московская область    Московская область      Московская область   
1137      Московская область    Московская область      Московская область   
1140                  Москва                Москва                  Москва   
1147                  Москва                Москва                  Москва   
1167      Московская область    Московская область      Московская область   
1191      Московская область    Московская область      Московская область   
1207      Московская область    Московская область      Московская область   
1232      Московская область    Московская область      Московская область   
1347      Московская область    Московская область      Московская область   
1404      Московская область    Московская область      Московская область   
1458      Московская область    Московская область      Московская область   
1468      Московская область    Московская область      Московская область   
1548   Архангельская область                Москва   Архангельская область   
1566      Московская область    Московская область      Московская область   
1609      Московская область    Московская область      Московская область   
...                      ...                   ...                     ...   
28120     Московская область    Московская область      Московская область   
28172     Московская область    Московская область      Московская область   
28263     Ростовская область    Московская область      Московская область   
28564                 Москва                Москва                  Москва   
28565     Кабардино-Балкария                Москва                  Москва   
28570     Московская область    Московская область      Московская область   
28574     Московская область    Московская область      Московская область   
28575      Псковская область                Москва                  Москва   
28581     Московская область    Московская область      Московская область   
28586                 Москва                Москва                  Москва   
28590                 Москва              Дагестан                  Москва   
28599     Московская область    Московская область      Московская область   
28601     Московская область    Московская область      Московская область   
28653                 Москва                Москва                  Москва   
28659     Московская область    Московская область      Московская область   
28670     Московская область    Московская область      Московская область   
28676                 Москва                Москва                  Москва   
28781     Московская область    Московская область      Московская область   
28794      Тюменская область     Тюменская область       Тюменская область   
28950                 Москва  

In [32]:
#отдельно посмотрим есть ли в логической переменной REG_FACT_POST_TP_FL для TP_PROVINCE единицы

geo_df[geo_df['REG_FACT_POST_TP_FL'] == 1]

Empty DataFrame
Columns: [REG_ADDRESS_PROVINCE, FACT_ADDRESS_PROVINCE, POSTAL_ADDRESS_PROVINCE, TP_PROVINCE, REGION_NM, REG_FACT_FL, REG_POST_FL, REG_FACT_POST_FL, REG_FACT_POST_TP_FL]
Index: []

*Ответ на задание 1:* <br>
пропуски есть. основные группы/факторы/причины:<br>
1. __**проблема:**__ есть пропуски для неработающих пенсионеров в 6 столбцах (и совсем небольшое число пропусков там-же, которые можно списать на ошибки при сборе данных<br>
__**решение:**__ можно сделать отдельный "класс" - неработающий пенсионер
2. __**проблема:**__ 590 пропусков TP_PROVINCE - причина пока не ясна. <br>__**решение:**__ Но есть много других географических данных, можно заполнить данными из одного из них для каждого конкретного клиента. Делаем допущение, что часто они совпадают (нужно проверить!!)
3. __**проблема:**__ PREVIOUS_CARD_NUM_UTILIZED <br>__**решение:**__ согласно описанию данных: пропуск = 0. просто заполним

#### Задание 2. Есть ли категориальные признаки? Что с ними делать?

In [44]:
#чтобы найти категориальные признаки - посмотрим к-во уникальных значений в каждом столбце (без учета nan)
distinct_in_cols = {i : df[i].value_counts(dropna = True).count() for i in df.columns}
distinct_in_cols

{'AGE': 47,
 'AGREEMENT_RK': 30133,
 'AUTO_RUS_FL': 2,
 'CHILD_TOTAL': 10,
 'COT_PRESENCE_FL': 2,
 'CREDIT': 12911,
 'DEPENDANTS': 8,
 'DL_DOCUMENT_FL': 1,
 'EDUCATION': 7,
 'FACT_ADDRESS_PROVINCE': 83,
 'FACT_LIVING_TERM': 525,
 'FACT_PHONE_FL': 2,
 'FACT_POST_FL': 2,
 'FAMILY_INCOME': 5,
 'FL_PRESENCE_FL': 2,
 'FST_PAYMENT': 4655,
 'GAR_PRESENCE_FL': 2,
 'GENDER': 2,
 'GEN_INDUSTRY': 31,
 'GEN_PHONE_FL': 2,
 'GEN_TITLE': 12,
 'GPF_DOCUMENT_FL': 2,
 'HS_PRESENCE_FL': 2,
 'JOB_DIR': 10,
 'LAND_PRESENCE_FL': 2,
 'LOAN_AVG_DLQ_AMT': 1942,
 'LOAN_DLQ_NUM': 14,
 'LOAN_MAX_DLQ': 7,
 'LOAN_MAX_DLQ_AMT': 1669,
 'LOAN_NUM_CLOSED': 10,
 'LOAN_NUM_PAYM': 88,
 'LOAN_NUM_TOTAL': 9,
 'MARITAL_STATUS': 5,
 'ORG_TP_FCAPITAL': 2,
 'ORG_TP_STATE': 5,
 'OWN_AUTO': 3,
 'PERSONAL_INCOME': 373,
 'POSTAL_ADDRESS_PROVINCE': 82,
 'PREVIOUS_CARD_NUM_UTILIZED': 2,
 'REGION_NM': 11,
 'REG_ADDRESS_PROVINCE': 82,
 'REG_FACT_FL': 2,
 'REG_FACT_POST_FL': 2,
 'REG_FACT_POST_TP_FL': 2,
 'REG_PHONE_FL': 2,
 'REG_POST_F

In [ ]:
#1) в выдорке есть категориальные признаки, уже закодированные в приемлемом для обучения модели виде (1-0):
already_binary = ['AUTO_RUS_FL', 'COT_PRESENCE_FL', 'FL_PRESENCE_FL', 'GAR_PRESENCE_FL', 'GENDER', 'GEN_PHONE_FL',
                 'GEN_TITLE', 'GPF_DOCUMENT_FL', 'HS_PRESENCE_FL', 'LAND_PRESENCE_FL', 'ORG_TP_FCAPITAL',
                 'REG_FACT_FL', 'REG_FACT_POST_FL', 'REG_FACT_POST_TP_FL', 'REG_PHONE_FL', 'REG_POST_FL',
                 'SOCSTATUS_PENS_FL', 'SOCSTATUS_WORK_FL']

#2) список категориальных признаков:
category_vars = ['EDUCATION', 'FACT_PHONE_FL', 'FAMILY_INCOME', 'GEN_INDUSTRY', 'JOB_DIR', 'LOAN_MAX_DLQ',
                'MARITAL_STATUS','ORG_TP_STATE', 'POSTAL_ADDRESS_PROVINCE', 'REGION_NM', 'REG_ADDRESS_PROVINCE',
                'TP_PROVINCE', ]
#непонятно LOAN_MAX_DLQ - номер просрочки качественный или нет


#3) есть параметр DL_DOCUMENT_FL, который на всей выборке = 0. Удалим его
df.drop('DL_DOCUMENT_FL', axis=1, inplace=True)

#### Задание 3. Фунция предобработки

Напишите функцию, которая бы

* Удаляло идентификатор `AGREEMENT_RK`
* Избавлялась от проблем с '.' и ',' в стобцах PERSONAL_INCOME, CREDIT, FST_PAYMENT, LOAN_AVG_DLQ_AMT, LOAN_MAX_DLQ_AMT
* Что-то делала с пропусками
* Кодировала категориальные признаки

В результате, ваш датафрейм должен содержать только числа и не содержать пропусков!

In [ ]:
def preproc_data(df_input):
    df_output = df_input.copy()
    
    
    ## Your Code Here
    
    return df_output

In [ ]:
df_preproc = df.pipe(preproc_data)

df_train_preproc = df_preproc.query('sample == "train"').drop(['sample'], axis=1)
df_test_preproc = df_preproc.query('sample == "test"').drop(['sample'], axis=1)

#### Задание 4. Отделите целевую переменную и остальные признаки

Должно получится:
* 2 матрицы: X и X_test
* 2 вектора: y и y_test

#### Задание 5. Обучение и оценка качества разных моделей

In [ ]:
from sklearn.cross_validation import train_test_split
# test_size=0.3, random_state=42

## Your Code Here


In [ ]:
# Попробовать следующие "черные ящики": интерфейс одинаковый 
# Постепенно мы узнаем, что они делают а сейчас учимся понимать какой работает качественнее
#     fit, 
#     predict, 
#     predict_proba

from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression

## Your Code Here

In [ ]:
# Посчитать метрики стандартные
# accuracy, precision, recall

from sklearn.metrics import accuracy_score, precision_score, recall_score

## Your Code Here

In [ ]:
# Визуалищировать эти метрики всех моделей на одном графике (чтоб визуально сравнить)
# Возможно вас удивит качество! Но задача подобрана специально ;) Такое качество тоже бывает

## Your Code Here

In [ ]:
# Потроить roc-кривые всех можелей на одном графике
# Вывести roc_auc каждой моделе
# Подпишите оси и линии

## Your Code Here

In [ ]:
from sklearn.cross_validation import cross_val_score
# Сделать k-fold (10 фолдов) кросс-валидацию каждой модели
# И посчитать средний roc_auc

## Your Code Here

In [ ]:
# Взять лучшую модель и сделать предсказания (с вероятностями (!!!)) для test выборки

predict = 
## Your Code Here

In [ ]:
# Померить roc_auc на тесте
# Вывести текстом и на графике =)

## Your Code Here

### Больше обязательных заданий нет, но ниже вы можете провести исследование, поиграться с данными
### Это возможность написать код, который я посмотрю и в случае чего откомметирую. ;)
### Это не оценивается и остается на ваше усмотрение. Просто дополнительная возможность

In [ ]:
### И далее ;)